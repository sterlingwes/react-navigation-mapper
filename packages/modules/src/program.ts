import * as ts from "typescript";
import * as path from "path";
import { globSync } from "glob";

type FileLookup = {
  [fileName: string]: {
    dependsOn: [string, string][];
  };
};

interface DrilldownOptions {
  onSourceFile?: (sourceFile: ts.SourceFile) => void;
}

interface ModuleOptions extends DrilldownOptions {
  moduleSearchLocations: string[];
  rootPath: string;
  files: FileLookup;
}

interface ProgramLookupOptions extends DrilldownOptions {
  rootPath?: string;
}

export const createProgramLookup = (
  projectPath: string,
  options?: ProgramLookupOptions
) => {
  const files: FileLookup = {};
  const basePath = projectPath.startsWith("./")
    ? projectPath
    : `./${projectPath}`;
  const paths = globSync(`${basePath}/**/*.tsx`);
  const rootPath = (options?.rootPath ?? path.dirname(__dirname)) + "/";

  const program = compile(paths, {
    ...options,
    moduleSearchLocations: [basePath],
    rootPath,
    files,
  });
  return { program, files };
};

function compile(sourceFiles: string[], moduleOptions: ModuleOptions) {
  const compilerOptions: ts.CompilerOptions = {
    module: ts.ModuleKind.AMD,
    target: ts.ScriptTarget.ES5,
  };
  const host = createCompilerHost(compilerOptions, moduleOptions);
  const program = ts.createProgram(sourceFiles, compilerOptions, host);
  return program;
}

function createCompilerHost(
  compilerOptions: ts.CompilerOptions,
  moduleOptions: ModuleOptions
): ts.CompilerHost {
  return {
    getSourceFile: (fileName, languageVersion, onError) => {
      const sourceFile = getSourceFile(fileName, languageVersion, onError);
      if (sourceFile && moduleOptions.onSourceFile) {
        moduleOptions.onSourceFile(sourceFile);
      }
      return sourceFile;
    },
    getDefaultLibFileName: () => "lib.d.ts",
    writeFile: (fileName, content) => ts.sys.writeFile(fileName, content),
    getCurrentDirectory: () => ts.sys.getCurrentDirectory(),
    getDirectories: (path) => ts.sys.getDirectories(path),
    getCanonicalFileName: (fileName) =>
      ts.sys.useCaseSensitiveFileNames ? fileName : fileName.toLowerCase(),
    getNewLine: () => ts.sys.newLine,
    useCaseSensitiveFileNames: () => ts.sys.useCaseSensitiveFileNames,
    fileExists,
    readFile,
    resolveModuleNames,
  };

  function fileExists(fileName: string): boolean {
    return ts.sys.fileExists(fileName);
  }

  function readFile(fileName: string): string | undefined {
    return ts.sys.readFile(fileName);
  }

  function getSourceFile(
    fileName: string,
    languageVersion: ts.ScriptTarget | ts.CreateSourceFileOptions,
    onError?: (message: string) => void
  ) {
    const sourceText = ts.sys.readFile(fileName);
    return sourceText !== undefined
      ? ts.createSourceFile(fileName, sourceText, languageVersion)
      : undefined;
  }

  function resolveModuleNames(
    moduleNames: string[],
    containingFile: string
  ): ts.ResolvedModule[] {
    const resolvedModules: ts.ResolvedModule[] = [];
    const containingFileLocalPath = containingFile.replace(
      moduleOptions.rootPath,
      ""
    );

    let file = moduleOptions.files[containingFileLocalPath];
    if (!file) {
      file = moduleOptions.files[containingFileLocalPath] = {
        dependsOn: [],
      };
    }

    for (const moduleName of moduleNames) {
      // try to use standard resolution
      let result = ts.resolveModuleName(
        moduleName,
        containingFile,
        compilerOptions,
        {
          fileExists,
          readFile,
        }
      );
      if (result.resolvedModule) {
        file.dependsOn.push([
          moduleName,
          result.resolvedModule.resolvedFileName.replace(
            moduleOptions.rootPath,
            ""
          ),
        ]);
        resolvedModules.push(result.resolvedModule);
      } else {
        // check fallback locations, for simplicity assume that module at location
        // should be represented by '.d.ts' file
        for (const location of moduleOptions.moduleSearchLocations) {
          const modulePath = path.join(location, moduleName + ".d.ts");
          if (fileExists(modulePath)) {
            resolvedModules.push({ resolvedFileName: modulePath });
          } else {
            // TEST
            resolvedModules.push({
              resolvedFileName: `unmatched-${moduleName}.ts`,
            });
          }
        }
      }
    }
    return resolvedModules;
  }
}
