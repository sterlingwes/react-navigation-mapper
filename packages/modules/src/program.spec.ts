import * as path from "path";
import { describe, it, expect } from "bun:test";
import { createProgramLookup } from "./program";

const rootPath = path.resolve(__dirname, "../../../");

describe("createProgramLookup", () => {
  it("should return the program and files lookup", () => {
    const { program, files } = createProgramLookup(
      "submodules/react-navigation/example/src",
      { rootPath }
    );
    expect(files).toMatchSnapshot();
  });

  it('should callback with source files with "onSourceFile" option', () => {
    const callbacks: string[] = [];
    createProgramLookup("submodules/react-navigation/example/src", {
      rootPath,
      onSourceFile: (sourceFile) => {
        callbacks.push(sourceFile.fileName);
      },
    });
    expect(callbacks).toMatchSnapshot();
  });
});
