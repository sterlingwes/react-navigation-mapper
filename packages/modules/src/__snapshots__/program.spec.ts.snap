// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`createProgramLookup should return the program and files lookup 1`] = `
{
  "submodules/react-navigation/example/src/Restart.native.tsx": {
    "dependsOn": [],
  },
  "submodules/react-navigation/example/src/Screens/AuthFlow.tsx": {
    "dependsOn": [],
  },
  "submodules/react-navigation/example/src/Screens/BottomTabs.tsx": {
    "dependsOn": [
      [
        "../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../Shared/Chat",
        "submodules/react-navigation/example/src/Shared/Chat.tsx",
      ],
      [
        "../Shared/Contacts",
        "submodules/react-navigation/example/src/Shared/Contacts.tsx",
      ],
      [
        "./SimpleStack",
        "submodules/react-navigation/example/src/Screens/SimpleStack.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/DrawerView.tsx": {
    "dependsOn": [],
  },
  "submodules/react-navigation/example/src/Screens/DynamicTabs.tsx": {
    "dependsOn": [],
  },
  "submodules/react-navigation/example/src/Screens/LinkComponent.tsx": {
    "dependsOn": [
      [
        "../screens",
        "submodules/react-navigation/example/src/screens.tsx",
      ],
      [
        "../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/MasterDetail.tsx": {
    "dependsOn": [
      [
        "../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
      [
        "../Shared/NewsFeed",
        "submodules/react-navigation/example/src/Shared/NewsFeed.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/MaterialBottomTabs.tsx": {
    "dependsOn": [
      [
        "../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../Shared/Chat",
        "submodules/react-navigation/example/src/Shared/Chat.tsx",
      ],
      [
        "../Shared/Contacts",
        "submodules/react-navigation/example/src/Shared/Contacts.tsx",
      ],
      [
        "./SimpleStack",
        "submodules/react-navigation/example/src/Screens/SimpleStack.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/MaterialTopTabs.tsx": {
    "dependsOn": [
      [
        "../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../Shared/Chat",
        "submodules/react-navigation/example/src/Shared/Chat.tsx",
      ],
      [
        "../Shared/Contacts",
        "submodules/react-navigation/example/src/Shared/Contacts.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/MixedHeaderMode.tsx": {
    "dependsOn": [
      [
        "../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
      [
        "../Shared/NewsFeed",
        "submodules/react-navigation/example/src/Shared/NewsFeed.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/MixedStack.tsx": {
    "dependsOn": [
      [
        "../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/ModalStack.tsx": {
    "dependsOn": [
      [
        "../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/NativeStack.tsx": {
    "dependsOn": [
      [
        "../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
      [
        "../Shared/NewsFeed",
        "submodules/react-navigation/example/src/Shared/NewsFeed.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/NativeStackHeaderCustomization.tsx": {
    "dependsOn": [
      [
        "../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
      [
        "../Shared/NewsFeed",
        "submodules/react-navigation/example/src/Shared/NewsFeed.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/NativeStackPreventRemove.tsx": {
    "dependsOn": [
      [
        "../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/NotFound.tsx": {
    "dependsOn": [
      [
        "../screens",
        "submodules/react-navigation/example/src/screens.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/SimpleStack.tsx": {
    "dependsOn": [
      [
        "../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
      [
        "../Shared/NewsFeed",
        "submodules/react-navigation/example/src/Shared/NewsFeed.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/StackHeaderCustomization.tsx": {
    "dependsOn": [
      [
        "../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
      [
        "../Shared/BlurView",
        "submodules/react-navigation/example/src/Shared/BlurView.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/StackPreventRemove.tsx": {
    "dependsOn": [
      [
        "../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/StackTransparent.tsx": {
    "dependsOn": [
      [
        "../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
      [
        "../Shared/NewsFeed",
        "submodules/react-navigation/example/src/Shared/NewsFeed.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/TabView.tsx": {
    "dependsOn": [
      [
        "./TabView/AutoWidthTabBar",
        "submodules/react-navigation/example/src/Screens/TabView/AutoWidthTabBar.tsx",
      ],
      [
        "./TabView/Coverflow",
        "submodules/react-navigation/example/src/Screens/TabView/Coverflow.tsx",
      ],
      [
        "./TabView/CustomIndicator",
        "submodules/react-navigation/example/src/Screens/TabView/CustomIndicator.tsx",
      ],
      [
        "./TabView/CustomTabBar",
        "submodules/react-navigation/example/src/Screens/TabView/CustomTabBar.tsx",
      ],
      [
        "./TabView/ScrollableTabBar",
        "submodules/react-navigation/example/src/Screens/TabView/ScrollableTabBar.tsx",
      ],
      [
        "./TabView/TabBarIcon",
        "submodules/react-navigation/example/src/Screens/TabView/TabBarIcon.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/TabView/AutoWidthTabBar.tsx": {
    "dependsOn": [
      [
        "../../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
      [
        "../../Shared/Chat",
        "submodules/react-navigation/example/src/Shared/Chat.tsx",
      ],
      [
        "../../Shared/Contacts",
        "submodules/react-navigation/example/src/Shared/Contacts.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/TabView/Coverflow.tsx": {
    "dependsOn": [],
  },
  "submodules/react-navigation/example/src/Screens/TabView/CustomIndicator.tsx": {
    "dependsOn": [
      [
        "../../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
      [
        "../../Shared/Contacts",
        "submodules/react-navigation/example/src/Shared/Contacts.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/TabView/CustomTabBar.tsx": {
    "dependsOn": [
      [
        "../../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
      [
        "../../Shared/Chat",
        "submodules/react-navigation/example/src/Shared/Chat.tsx",
      ],
      [
        "../../Shared/Contacts",
        "submodules/react-navigation/example/src/Shared/Contacts.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/TabView/ScrollableTabBar.tsx": {
    "dependsOn": [
      [
        "../../Shared/Albums",
        "submodules/react-navigation/example/src/Shared/Albums.tsx",
      ],
      [
        "../../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
      [
        "../../Shared/Chat",
        "submodules/react-navigation/example/src/Shared/Chat.tsx",
      ],
      [
        "../../Shared/Contacts",
        "submodules/react-navigation/example/src/Shared/Contacts.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Screens/TabView/TabBarIcon.tsx": {
    "dependsOn": [
      [
        "../../Shared/Article",
        "submodules/react-navigation/example/src/Shared/Article.tsx",
      ],
      [
        "../../Shared/Chat",
        "submodules/react-navigation/example/src/Shared/Chat.tsx",
      ],
      [
        "../../Shared/Contacts",
        "submodules/react-navigation/example/src/Shared/Contacts.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/Shared/Albums.tsx": {
    "dependsOn": [],
  },
  "submodules/react-navigation/example/src/Shared/Article.tsx": {
    "dependsOn": [],
  },
  "submodules/react-navigation/example/src/Shared/BlurView.native.tsx": {
    "dependsOn": [],
  },
  "submodules/react-navigation/example/src/Shared/BlurView.tsx": {
    "dependsOn": [],
  },
  "submodules/react-navigation/example/src/Shared/Chat.tsx": {
    "dependsOn": [],
  },
  "submodules/react-navigation/example/src/Shared/Contacts.tsx": {
    "dependsOn": [],
  },
  "submodules/react-navigation/example/src/Shared/NewsFeed.tsx": {
    "dependsOn": [],
  },
  "submodules/react-navigation/example/src/Shared/SettingsItem.tsx": {
    "dependsOn": [],
  },
  "submodules/react-navigation/example/src/index.tsx": {
    "dependsOn": [
      [
        "./Restart",
        "submodules/react-navigation/example/src/Restart.tsx",
      ],
      [
        "./screens",
        "submodules/react-navigation/example/src/screens.tsx",
      ],
      [
        "./Screens/NotFound",
        "submodules/react-navigation/example/src/Screens/NotFound.tsx",
      ],
      [
        "./Shared/SettingsItem",
        "submodules/react-navigation/example/src/Shared/SettingsItem.tsx",
      ],
    ],
  },
  "submodules/react-navigation/example/src/screens.tsx": {
    "dependsOn": [
      [
        "./Screens/AuthFlow",
        "submodules/react-navigation/example/src/Screens/AuthFlow.tsx",
      ],
      [
        "./Screens/BottomTabs",
        "submodules/react-navigation/example/src/Screens/BottomTabs.tsx",
      ],
      [
        "./Screens/DrawerView",
        "submodules/react-navigation/example/src/Screens/DrawerView.tsx",
      ],
      [
        "./Screens/DynamicTabs",
        "submodules/react-navigation/example/src/Screens/DynamicTabs.tsx",
      ],
      [
        "./Screens/LinkComponent",
        "submodules/react-navigation/example/src/Screens/LinkComponent.tsx",
      ],
      [
        "./Screens/MasterDetail",
        "submodules/react-navigation/example/src/Screens/MasterDetail.tsx",
      ],
      [
        "./Screens/MaterialBottomTabs",
        "submodules/react-navigation/example/src/Screens/MaterialBottomTabs.tsx",
      ],
      [
        "./Screens/MaterialTopTabs",
        "submodules/react-navigation/example/src/Screens/MaterialTopTabs.tsx",
      ],
      [
        "./Screens/MixedHeaderMode",
        "submodules/react-navigation/example/src/Screens/MixedHeaderMode.tsx",
      ],
      [
        "./Screens/MixedStack",
        "submodules/react-navigation/example/src/Screens/MixedStack.tsx",
      ],
      [
        "./Screens/ModalStack",
        "submodules/react-navigation/example/src/Screens/ModalStack.tsx",
      ],
      [
        "./Screens/NativeStack",
        "submodules/react-navigation/example/src/Screens/NativeStack.tsx",
      ],
      [
        "./Screens/NativeStackHeaderCustomization",
        "submodules/react-navigation/example/src/Screens/NativeStackHeaderCustomization.tsx",
      ],
      [
        "./Screens/NativeStackPreventRemove",
        "submodules/react-navigation/example/src/Screens/NativeStackPreventRemove.tsx",
      ],
      [
        "./Screens/SimpleStack",
        "submodules/react-navigation/example/src/Screens/SimpleStack.tsx",
      ],
      [
        "./Screens/StackHeaderCustomization",
        "submodules/react-navigation/example/src/Screens/StackHeaderCustomization.tsx",
      ],
      [
        "./Screens/StackPreventRemove",
        "submodules/react-navigation/example/src/Screens/StackPreventRemove.tsx",
      ],
      [
        "./Screens/StackTransparent",
        "submodules/react-navigation/example/src/Screens/StackTransparent.tsx",
      ],
      [
        "./Screens/TabView",
        "submodules/react-navigation/example/src/Screens/TabView.tsx",
      ],
    ],
  },
}
`;

exports[`createProgramLookup should callback with source files with "onSourceFile" option 1`] = `
[
  "submodules/react-navigation/example/src/Restart.native.tsx",
  "submodules/react-navigation/example/src/Restart.tsx",
  "submodules/react-navigation/example/src/screens.tsx",
  "submodules/react-navigation/example/src/index.tsx",
  "submodules/react-navigation/example/src/Screens/NotFound.tsx",
  "submodules/react-navigation/example/src/Screens/NativeStackHeaderCustomization.tsx",
  "submodules/react-navigation/example/src/Screens/MaterialTopTabs.tsx",
  "submodules/react-navigation/example/src/Screens/MixedStack.tsx",
  "submodules/react-navigation/example/src/Screens/ModalStack.tsx",
  "submodules/react-navigation/example/src/Screens/NativeStack.tsx",
  "submodules/react-navigation/example/src/Screens/StackTransparent.tsx",
  "submodules/react-navigation/example/src/Screens/MaterialBottomTabs.tsx",
  "submodules/react-navigation/example/src/Screens/DrawerView.tsx",
  "submodules/react-navigation/example/src/Screens/BottomTabs.tsx",
  "submodules/react-navigation/example/src/Screens/NativeStackPreventRemove.tsx",
  "submodules/react-navigation/example/src/Screens/StackPreventRemove.tsx",
  "submodules/react-navigation/example/src/Screens/MixedHeaderMode.tsx",
  "submodules/react-navigation/example/src/Screens/LinkComponent.tsx",
  "submodules/react-navigation/example/src/Screens/SimpleStack.tsx",
  "submodules/react-navigation/example/src/Screens/TabView.tsx",
  "submodules/react-navigation/example/src/Screens/StackHeaderCustomization.tsx",
  "submodules/react-navigation/example/src/Screens/AuthFlow.tsx",
  "submodules/react-navigation/example/src/Screens/DynamicTabs.tsx",
  "submodules/react-navigation/example/src/Screens/MasterDetail.tsx",
  "submodules/react-navigation/example/src/Screens/TabView/CustomTabBar.tsx",
  "submodules/react-navigation/example/src/Screens/TabView/Coverflow.tsx",
  "submodules/react-navigation/example/src/Screens/TabView/ScrollableTabBar.tsx",
  "submodules/react-navigation/example/src/Screens/TabView/TabBarIcon.tsx",
  "submodules/react-navigation/example/src/Screens/TabView/AutoWidthTabBar.tsx",
  "submodules/react-navigation/example/src/Screens/TabView/CustomIndicator.tsx",
  "submodules/react-navigation/example/src/Shared/Contacts.tsx",
  "submodules/react-navigation/example/src/Shared/Albums.tsx",
  "submodules/react-navigation/example/src/Shared/Chat.tsx",
  "submodules/react-navigation/example/src/Shared/Article.tsx",
  "submodules/react-navigation/example/src/Shared/BlurView.native.tsx",
  "submodules/react-navigation/example/src/Shared/SettingsItem.tsx",
  "submodules/react-navigation/example/src/Shared/NewsFeed.tsx",
  "submodules/react-navigation/example/src/Shared/BlurView.tsx",
]
`;
